Add a unit test for missing exec event in the eventcache

tpapagian opened the pull request:
This unit-test emulates the case where we get an exit event but we miss the exec event. In that case, we should wait for the cache iterations to end and then produce an exit event with partial process info. In that case we only have the process start time and the process pid.

This PR also fixes https://github.com/cilium/tetragon/issues/285 and https://github.com/cilium/tetragon/issues/247. OOO events was related to these issues and thus I don't use a separate PR for them. To test this I have run 8 successful consecutive runs in the CI while before a kprobe flake appears every ~3 runs. 

Signed-off-by: Anastasios Papagiannis <tasos.papagiannnis@gmail.com>

willfindlay commented on pkg/api/readyapi/readyapi.go:
```diff
@@ -9,3 +9,9 @@ type MsgTETRAGONReady struct{}
 	return nil
 }
+
+func (msg *MsgTETRAGONReady) SetCached() {}
```
Maybe a comment here to explain that MessageReady.IsCached() should always be true and that's why we're leaving this blank?

willfindlay reviewed (approved):
Just a minor comment from me.

jrfastab commented:
I'm somehow missing how a msg in the eventcache gets 'readded' to the cache. The eventcache kicks the event out to the NotifyListeners these can do an Add() in the eventcache again?

jrfastab commented:
Aha so fairly sure the issue here is the dummyInterface is calling HandleMessage which does not reflect what the actual code does run from main does. The flow is `HandleMessage -> Exporter` with some constructors through processManager to add the exporter as a listener and then server.go used to construct the exporter correctly. To follow along the relevant files are:

 pkg/exporter/exporter.go: Implements the Send() method and Start() methods to satisfy the interface
 pkg/server/server.go: Embeds into exporter above and used by exporter to do aggregation and filtering
 pkg/grpc//process_manager.go: Has the Add logic to add multiple listeners. This lets us run both gRPC listeners, dbg listeners and json writers.

jrfastab reviewed (changes requested):
Fixing dummy notifier should resolve issue and then we don't need patch 3.

kkourt commented on pkg/eventcache/eventcache.go:
```diff
@@ -19,14 +19,11 @@ import (
-// garbage collection run interval
-const (
-	eventRetryTimer = time.Second * 10
+	// garbage collection states
```
nit: is states a typo?

kkourt commented on pkg/eventcache/eventcache.go:
```diff
@@ -21,9 +21,9 @@ import (
 
```
>  Make eventRetryTimer and cacheStrikes from eventcache public 

Maybe add something like:

> 
> We do this so that we can use the value in the json checker and in testing. See next two commit for details.

in the commit message?

kkourt reviewed (approved):
Thanks! 

Nice catch on the timing issue. I believe this will fix the linked issues.

tpapagian commented:
> Nice catch on the timing issue. I believe this will fix the linked issues.

@kkourt It seems that it still have some flakes related to these issues. This makes things better but I believe that it does not solve this entirely. Working on that (this is the reason that this PR is in draft).

willfindlay commented on pkg/eventcache/eventcache.go:
```diff
@@ -99,20 +98,25 @@ func (ec *Cache) handleEvents() {
 		}
 
 		if err != nil {
+			logger.GetLogger().WithError(err).Warnf("handleEvents() - %d - %d", callFn, event.color)
```
I think this is going to spam our logs with a lot of messages, since we expect err to not be nil quite frequently. I would suggest moving it to Debugf

tpapagian replied:
Just trying to debug the issue here. I don't plan to keep them when the issue is resolved. (Same for the other log messages)

willfindlay commented on pkg/process/cache.go:
```diff
@@ -112,6 +112,7 @@ func (pc *Cache) cacheGarbageCollector() {
 }
 
 func (pc *Cache) deletePending(process *ProcessInternal) {
+	logger.GetLogger().Warnf("DELETE EXEC - PID: %d - TIME: %d - EXECID: %s", process.process.Pid.Value, process.process.ExecId)
```
Do we expect this to happen frequently? If so let's move it to Debugf imo

willfindlay commented on pkg/process/cache.go:
```diff
@@ -127,6 +128,7 @@ func (pc *Cache) refInc(p *ProcessInternal) {
 }
 
 func (pc *Cache) Purge() {
+	logger.GetLogger().Warnf("procCache Purge")
```
Same question as above

willfindlay commented on pkg/eventcache/eventcache.go:
```diff
@@ -80,6 +78,7 @@ func handleEvent(event *cacheObj) error {
 	// now because it should be available. Otherwise we have a valid
 	// process and lets copy it across.
 	if event.internal == nil {
+		logger.GetLogger().Warnf("handleEvent() - PID: %d - TIME: %d - EXECID: %s", p.Pid.Value, event.timestamp, process.GetProcessID(p.Pid.Value, event.timestamp))
```
Do we expect this to happen frequently? If so let's move it to Debugf imo

tpapagian commented:
> Fixing dummy notifier should resolve issue and then we don't need patch 3.

Fixed dummy notifier and addressed the other comments. Ready for review now.

jrfastab commented on pkg/eventcache/eventcache.go:
```diff
@@ -19,14 +19,11 @@ import (
+	// garbage collection retries
+	cacheStrikes = 15
+	// garbage collection run interval
+	eventRetryTimer = time.Second * 2
```
The 10seconds was about balancing GC performance cost vs getting events out. We should add a configMap option for this I think.

tpapagian replied:
Created https://github.com/cilium/tetragon/issues/296 to keep track of that.

jrfastab commented:
This all looks good to me. With one question about making the time to retry cache configurable. For testing and watching results over grpc or cmd line its nice to have the system be more reactive, but when writing to logs for offsystem analysis I'm not sure it matters much if its 2 seconds or 2minutes.
